Source/WebCore/html/canvas/CanvasRenderingContext2DBase.cpp

```diff
@@ -2309,6 +2320,92 @@ ExceptionOr<Ref<ImageData>> CanvasRenderingContext2DBase::createImageData(int sw
     return imageData;
 }
 
+static constexpr unsigned minimumConsecutiveCachedImageDataRequestsBeforeCaching = 2;
+static constexpr unsigned imageDataSizeThresholdForCaching = 60 * 60;
+
+bool CanvasRenderingContext2DBase::cacheImageDataIfPossible(ImageData& data, const IntPoint& destinationPosition, const IntRect& sourceRect)
+{
+    if (!destinationPosition.isZero() || !sourceRect.location().isZero() || sourceRect.size() != data.size() || sourceRect.size() != canvasBase().size())
+        return false;
+
+    if (data.colorSpace() != m_settings.colorSpace)
+        return false;
+
+    if (data.size().area() > imageDataSizeThresholdForCaching)
+        return false;
+
+    if (m_cachedImageData) {
+        if (++m_cachedImageData->requestCount >= minimumConsecutiveCachedImageDataRequestsBeforeCaching) {
+            m_cachedImageData->imageData = data.clone();
+            m_cachedImageData->ageOfCache = 0;
+
+            ConstPixelBufferConversionView source {
+                .format = {
+                    .alphaFormat = AlphaPremultiplication::Unpremultiplied,
+                    .pixelFormat = PixelFormat::RGBA8,
+                    .colorSpace = DestinationColorSpace::SRGB(),
+                },
+                .bytesPerRow = static_cast<unsigned>(data.size().width() * 4),
+                .rows = m_cachedImageData->imageData->data().data(),
+            };
+            PixelBufferConversionView destination {
+                .format = {
+                    .alphaFormat = AlphaPremultiplication::Premultiplied,
+                    .pixelFormat = PixelFormat::RGBA8,
+                    .colorSpace = DestinationColorSpace::SRGB(),
+                },
+                .bytesPerRow = static_cast<unsigned>(data.size().width() * 4),
+                .rows = m_cachedImageData->imageData->data().data(),
+            };
+            convertImagePixels(source, destination, data.size());
```

Comment:
Can we do this only when retrieving the ImageData from the cache? For cases where the cached ImageData is not re-used, it would be good to avoid the unnecessary conversion.